IndexOutOfBoundsException when JPQL has embeddable or relational attributes without the optional entity identification variable

JPQL queries with embeddable (or relation) attributes that omit the optional entity identification variable (for consistency with JDQL) get IndexOutOfBoundsException on EclipseLink.

Example query:
```
FROM Business WHERE location.address.city=?1 ORDER BY name
```

Here is the exception stack with `IndexOutOfBoundsException`:
```
[6/27/24, 13:26:52:316 CDT] 00000051 id=079f44f4 io.openliberty.data.internal.persistence.QueryInfo           < withJPQL Exit  
                                                                                                               QueryInfo@2b93bafe Page locatedIn(String, PageRequest, Order)  FROM Business WHERE location.address.city=?1 ORDER BY name [1 positional params]
[6/27/24, 13:26:52:338 CDT] 00000051 id=00000000 io.openliberty.data.internal.persistence.PageImpl            > <init> Entry  
                                                                                                               QueryInfo@2b93bafe Page locatedIn(String, PageRequest, Order)  FROM Business WHERE location.address.city=?1 ORDER BY name [1 positional params]
                                                                                                               PageRequest{page=1, size=6, mode=OFFSET}
                                                                                                               Rochester
                                                                                                               PageRequest{page=1, size=6, mode=OFFSET}
                                                                                                               [Sort[property=name, isAscending=true, ignoreCase=false]]
[6/27/24, 13:26:52:338 CDT] 00000051 id=7302420d enliberty.data.internal.persistence.service.DBStoreEMBuilder > createEntityManager Entry 
[6/27/24, 13:26:52:338 CDT] 00000051 id=3ea89612 om.ibm.wsspi.persistence.internal.PersistenceServiceUnitImpl > createEntityManager Entry 
[6/27/24, 13:26:52:339 CDT] 00000051 id=3ea89612 om.ibm.wsspi.persistence.internal.PersistenceServiceUnitImpl < createEntityManager Exit  
                                                                                                               org.eclipse.persistence.internal.jpa.EntityManagerImpl@ab3b5af
[6/27/24, 13:26:52:339 CDT] 00000051 id=7302420d enliberty.data.internal.persistence.service.DBStoreEMBuilder < createEntityManager Exit  
                                                                                                               org.eclipse.persistence.internal.jpa.EntityManagerImpl@ab3b5af
[6/27/24, 13:26:52:339 CDT] 00000051 id=00000000 io.openliberty.data.internal.persistence.RepositoryImpl      3 java.lang.IllegalArgumentException replaced with jakarta.data.exceptions.MappingException
[6/27/24, 13:26:52:339 CDT] 00000051 id=00000000 io.openliberty.data.internal.persistence.RepositoryImpl      3 Failure occurred: jakarta.data.exceptions.MappingException
[6/27/24, 13:26:52:339 CDT] 00000051 id=15d7a454 io.openliberty.data.internal.persistence.RepositoryImpl      < invoke MixedRepository.locatedIn Exit  
                                                                                                               jakarta.data.exceptions.MappingException: java.lang.IllegalArgumentException: An exception occurred while creating a query in EntityManager: 
Exception Description: Internal problem encountered while compiling [ FROM Business WHERE location.address.city=?1 ORDER BY name].
Internal Exception: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
	at io.openliberty.data.internal.persistence.RepositoryImpl.failure(RepositoryImpl.java:302)
	at io.openliberty.data.internal.persistence.PageImpl.<init>(PageImpl.java:67)
	at io.openliberty.data.internal.persistence.RepositoryImpl.invoke(RepositoryImpl.java:858)
	at jdk.proxy11/jdk.proxy11.$Proxy84.locatedIn(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.weld.bean.proxy.AbstractBeanInstance.invoke(AbstractBeanInstance.java:38)
	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:109)
	at test.jakarta.data.jpa.web.MixedRepository$1687930109$Proxy$_$$_WeldClientProxy.locatedIn(Unknown Source)
	at test.jakarta.data.jpa.web.DataJPATestServlet.testCountQueryWithFromAndWhereClausesOnly(DataJPATestServlet.java:496)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at componenttest.app.FATServlet.doGet(FATServlet.java:74)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:633)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:723)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1266)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:754)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:451)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1362)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1078)
	at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:77)
	at com.ibm.ws.webcontainer40.servlet.CacheServletWrapper40.handleRequest(CacheServletWrapper40.java:87)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:978)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:293)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1260)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:476)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:435)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:569)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:503)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:363)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:330)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:169)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:77)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:516)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:586)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:970)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1059)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:280)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalArgumentException: An exception occurred while creating a query in EntityManager: 
Exception Description: Internal problem encountered while compiling [ FROM Business WHERE location.address.city=?1 ORDER BY name].
Internal Exception: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.createQuery(EntityManagerImpl.java:1848)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.createQuery(EntityManagerImpl.java:1869)
	at io.openliberty.data.internal.persistence.PageImpl.<init>(PageImpl.java:58)
	... 39 more
Caused by: Exception [EclipseLink-0] (Eclipse Persistence Services - 5.0.0-B02.v202404111748): org.eclipse.persistence.exceptions.JPQLException
Exception Description: Internal problem encountered while compiling [ FROM Business WHERE location.address.city=?1 ORDER BY name].
Internal Exception: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
	at org.eclipse.persistence.internal.jpa.jpql.HermesParser.buildUnexpectedException(HermesParser.java:215)
	at org.eclipse.persistence.internal.jpa.jpql.HermesParser.populateQueryImp(HermesParser.java:310)
	at org.eclipse.persistence.internal.jpa.jpql.HermesParser.buildQuery(HermesParser.java:174)
	at org.eclipse.persistence.internal.jpa.EJBQueryImpl.buildEJBQLDatabaseQuery(EJBQueryImpl.java:144)
	at org.eclipse.persistence.internal.jpa.EJBQueryImpl.buildEJBQLDatabaseQuery(EJBQueryImpl.java:120)
	at org.eclipse.persistence.internal.jpa.EJBQueryImpl.<init>(EJBQueryImpl.java:107)
	at org.eclipse.persistence.internal.jpa.EJBQueryImpl.<init>(EJBQueryImpl.java:91)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.createQuery(EntityManagerImpl.java:1846)
	... 41 more
Caused by: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:385)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at org.eclipse.persistence.jpa.jpql.parser.AbstractPathExpression.getPath(AbstractPathExpression.java:258)
	at org.eclipse.persistence.internal.jpa.jpql.ExpressionBuilderVisitor$PathResolver.resolvePath(ExpressionBuilderVisitor.java:2476)
	at org.eclipse.persistence.internal.jpa.jpql.ExpressionBuilderVisitor$PathResolver.visitPathExpression(ExpressionBuilderVisitor.java:2675)
	at org.eclipse.persistence.internal.jpa.jpql.ExpressionBuilderVisitor$PathResolver.visit(ExpressionBuilderVisitor.java:2620)
	at org.eclipse.persistence.jpa.jpql.parser.IdentificationVariable.accept(IdentificationVariable.java:120)
	at org.eclipse.persistence.internal.jpa.jpql.ExpressionBuilderVisitor$PathResolver.visitPathExpression(ExpressionBuilderVisitor.java:2650)
	at org.eclipse.persistence.internal.jpa.jpql.ExpressionBuilderVisitor$PathResolver.visit(ExpressionBuilderVisitor.java:2620)
	at org.eclipse.persistence.jpa.jpql.parser.StateFieldPathExpression.accept(StateFieldPathExpression.java:77)
	at org.eclipse.persistence.internal.jpa.jpql.ExpressionBuilderVisitor.visitPathExpression(ExpressionBuilderVisitor.java:2127)
	at org.eclipse.persistence.internal.jpa.jpql.ExpressionBuilderVisitor.visit(ExpressionBuilderVisitor.java:1893)
	at org.eclipse.persistence.jpa.jpql.parser.StateFieldPathExpression.accept(StateFieldPathExpression.java:77)
	at org.eclipse.persistence.internal.jpa.jpql.ExpressionBuilderVisitor$ComparisonExpressionVisitor.visit(ExpressionBuilderVisitor.java:2208)
	at org.eclipse.persistence.jpa.jpql.parser.AnonymousExpressionVisitor.visit(AnonymousExpressionVisitor.java:466)
	at org.eclipse.persistence.jpa.jpql.parser.StateFieldPathExpression.accept(StateFieldPathExpression.java:77)
	at org.eclipse.persistence.internal.jpa.jpql.ExpressionBuilderVisitor.visit(ExpressionBuilderVisitor.java:732)
	at org.eclipse.persistence.jpa.jpql.parser.ComparisonExpression.accept(ComparisonExpression.java:71)
	at org.eclipse.persistence.internal.jpa.jpql.ExpressionBuilderVisitor.visit(ExpressionBuilderVisitor.java:2111)
	at org.eclipse.persistence.jpa.jpql.parser.WhereClause.accept(WhereClause.java:56)
	at org.eclipse.persistence.internal.jpa.jpql.ExpressionBuilderVisitor.buildExpression(ExpressionBuilderVisitor.java:280)
	at org.eclipse.persistence.internal.jpa.jpql.JPQLQueryContext.buildExpression(JPQLQueryContext.java:326)
	at org.eclipse.persistence.internal.jpa.jpql.AbstractObjectLevelReadQueryVisitor.visit(AbstractObjectLevelReadQueryVisitor.java:226)
	at org.eclipse.persistence.jpa.jpql.parser.WhereClause.accept(WhereClause.java:56)
	at org.eclipse.persistence.internal.jpa.jpql.AbstractObjectLevelReadQueryVisitor.visitAbstractSelectStatement(AbstractObjectLevelReadQueryVisitor.java:292)
	at org.eclipse.persistence.internal.jpa.jpql.AbstractObjectLevelReadQueryVisitor.visit(AbstractObjectLevelReadQueryVisitor.java:160)
	at org.eclipse.persistence.jpa.jpql.parser.SelectStatement.accept(SelectStatement.java:100)
	at org.eclipse.persistence.internal.jpa.jpql.HermesParser$DatabaseQueryVisitor.visit(HermesParser.java:440)
	at org.eclipse.persistence.jpa.jpql.parser.SelectStatement.accept(SelectStatement.java:100)
	at org.eclipse.persistence.internal.jpa.jpql.HermesParser$DatabaseQueryVisitor.visit(HermesParser.java:423)
	at org.eclipse.persistence.jpa.jpql.parser.JPQLExpression.accept(JPQLExpression.java:186)
	at org.eclipse.persistence.internal.jpa.jpql.HermesParser.populateQueryImp(HermesParser.java:296)
	... 47 more
```

--- a/dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/QueryInfo.java
+++ b/dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/QueryInfo.java
@@ -2951,17 +2951,7 @@ public class QueryInfo {
                         entityVar = "this";
                         entityVar_ = "";
 
-                        // TODO remove this workaround for #28931 once fixed
-                        boolean insertEntityVar = !entityInfo.relationAttributeNames.isEmpty();
-                        if (insertEntityVar) {
-                            entityVar_ = entityVar + ".";
-                            StringBuilder q = new StringBuilder(ql.length() * 3 / 2) //
-                                            .append("DELETE FROM ").append(entityInfo.name) //
-                                            .append(' ').append(entityVar).append(" WHERE");
-                            appendWithIdentifierName(ql, startAt + 5, ql.length(), entityVar_, q);
-                            jpql = q.toString();
-                            // The following if block is not part of the workaround and is still needed:
-                        } else if (entityInfo.recordClass != null)
+                        if (entityInfo.recordClass != null)
                             // Switch from record name to entity name
                             jpql = new StringBuilder(ql.length() + 6) //
                                             .append("DELETE FROM ") //
@@ -3009,16 +2999,7 @@ public class QueryInfo {
                         entityVar = "this";
                         entityVar_ = "";
 
-                        // TODO remove this workaround for #28931 once fixed
-                        boolean insertEntityVar = !entityInfo.relationAttributeNames.isEmpty();
-                        if (insertEntityVar) {
-                            entityVar = "o";
-                            entityVar_ = "o.";
-                            StringBuilder q = new StringBuilder(ql.length() * 3 / 2) //
-                                            .append("UPDATE ").append(entityInfo.name).append(" o SET");
-                            appendWithIdentifierName(ql, startAt + 3, ql.length(), entityVar_, q);
-                            jpql = q.toString();
-                        } else if (entityName.length() != entityInfo.name.length() || entityName.indexOf(entityInfo.name) != 0)
+                        if (entityName.length() != entityInfo.name.length() || entityName.indexOf(entityInfo.name) != 0)
                             jpql = new StringBuilder(ql.length() * 3 / 2) //
                                             .append("UPDATE ").append(entityInfo.name).append(" SET") //
                                             .append(jpql.substring(startAt + 3, ql.length())) //
@@ -3229,7 +3210,7 @@ public class QueryInfo {
 
             boolean hasEntityVar = entityVar_.length() > 0;
 
-            // TODO remove this workaround for #28931 once fixed
+            // TODO remove this workaround for #30351 once fixed and run the _fat_jpa bucket to verify
             boolean insertEntityVar = entityVar_.length() == 0 && !entityInfo.relationAttributeNames.isEmpty();
             if (insertEntityVar)
                 entityVar_ = entityVar + ".";
@@ -3333,7 +3314,7 @@ public class QueryInfo {
             }
 
             if (whereLen > 0)
-                // TODO once fixed, test #28931 by adding: && !"this.".equalsIgnoreCase(entityVar_)
+                // TODO once fixed, test #30351 by adding: && !"this.".equalsIgnoreCase(entityVar_)
                 // and running DataJPATestServlet.testCountQueryWithFromAndWhereClausesOnly
                 if (insertEntityVar) {
                     q.append("WHERE");

--- a/dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/QueryInfo.java
+++ b/dev/io.openliberty.data.internal.persistence/src/io/openliberty/data/internal/persistence/QueryInfo.java
@@ -395,20 +395,17 @@ public class QueryInfo {
 
     /**
      * Adds Sort criteria to the end of the tracked list of sort criteria.
-     * For IdClass, adds all Id properties separately. TODO use id(this) instead once #30093 is fixed
      *
      * @param ignoreCase if ordering is to be independent of case.
-     * @param attribute  name of attribute (@OrderBy value or Sort property or parsed from OrderBy query-by-method).
+     * @param attribute  name of attribute (@OrderBy value, Sort property,
+     *                       or parsed from OrderBy method name keyword).
      * @param descending if ordering is to be in descending order
      */
     @Trivial
     private void addSort(boolean ignoreCase, String attribute, boolean descending) {
-        Set<String> names = entityInfo.idClassAttributeAccessors != null && ID.equalsIgnoreCase(attribute) //
-                        ? entityInfo.idClassAttributeAccessors.keySet() //
-                        : Set.of(attribute);
-
-        for (String name : names) {
-            name = getAttributeName(name, true);
+        if (entityInfo.idClassAttributeAccessors == null ||
+            !ID.equalsIgnoreCase(attribute)) {
+            String name = getAttributeName(attribute, true);
 
             sorts.add(ignoreCase ? //
                             descending ? //
@@ -417,7 +414,27 @@ public class QueryInfo {
                             descending ? //
                                             Sort.desc(name) : //
                                             Sort.asc(name));
+        } else {
+            // Expand ID(THIS) for composite IdClass into separate attributes
+            for (String name : entityInfo.idClassAttributeAccessors.keySet())
+                addSort(ignoreCase, name, descending);
         }
+    }
+
+    /**
+     * Appends the attribute name if it is a function ending with a ')' character.
+     * Otherwise appends the attribute name prefixed by the entity identifier
+     * variable.
+     *
+     * @param attrName attribute name, which might be represented as a function
+     *                     such as id(o) or version(o)
+     * @param q        string builder to which to append
+     */
+    @Trivial
+    private final void appendAttributeName(String attrName, StringBuilder q) {
+        if (attrName.charAt(attrName.length() - 1) != ')')
+            q.append(entityVar_);
+        q.append(attrName);
     }
 
     /**
@@ -474,9 +491,7 @@ public class QueryInfo {
 
                     if ("id".equalsIgnoreCase(str) && ql.regionMatches(true, i + 1, "(THIS)", 0, 6)) {
                         String name = getAttributeName(By.ID, true);
-                        if (name.charAt(name.length() - 1) != ')')
-                            q.append(o_);
-                        q.append(name);
+                        appendAttributeName(name, q);
                         i += 6;
                     } else if ("this".equalsIgnoreCase(str)
                                || getAttributeName(str, false) == null) {
@@ -1549,10 +1564,7 @@ public class QueryInfo {
         if (trimmed)
             attributeExpr.append("TRIM(");
 
-        String o_ = entityVar_;
-        if (name.charAt(name.length() - 1) != ')')
-            attributeExpr.append(o_);
-        attributeExpr.append(name);
+        appendAttributeName(name, attributeExpr);
 
         if (trimmed)
             attributeExpr.append(')');
@@ -1664,7 +1676,6 @@ public class QueryInfo {
         String paramPrefix = jpqlParamNames.isEmpty() ? "?" : ":cursor";
         StringBuilder a = fwd == null ? null : new StringBuilder(200).append(hasWhere ? " AND (" : " WHERE (");
         StringBuilder b = prev == null ? null : new StringBuilder(200).append(hasWhere ? " AND (" : " WHERE (");
-        String o_ = entityVar_;
         for (int i = 0; i < numSorts; i++) {
             if (a != null)
                 a.append(i == 0 ? "(" : " OR (");
@@ -1677,21 +1688,27 @@ public class QueryInfo {
                 boolean lower = sort.ignoreCase();
                 if (a != null)
                     if (lower) {
-                        a.append(s == 0 ? "LOWER(" : " AND LOWER(").append(o_).append(name).append(')');
+                        a.append(s == 0 ? "LOWER(" : " AND LOWER(");
+                        appendAttributeName(name, a);
+                        a.append(')');
                         a.append(s < i ? '=' : (asc ? '>' : '<'));
                         a.append("LOWER(").append(paramPrefix).append(jpqlParamCount + 1 + s).append(')');
                     } else {
-                        a.append(s == 0 ? "" : " AND ").append(o_).append(name);
+                        a.append(s == 0 ? "" : " AND ");
+                        appendAttributeName(name, a);
                         a.append(s < i ? '=' : (asc ? '>' : '<'));
                         a.append(paramPrefix).append(jpqlParamCount + 1 + s);
                     }
                 if (b != null)
                     if (lower) {
-                        b.append(s == 0 ? "LOWER(" : " AND LOWER(").append(o_).append(name).append(')');
+                        b.append(s == 0 ? "LOWER(" : " AND LOWER(");
+                        appendAttributeName(name, b);
+                        b.append(')');
                         b.append(s < i ? '=' : (asc ? '<' : '>'));
                         b.append("LOWER(").append(paramPrefix).append(jpqlParamCount + 1 + s).append(')');
                     } else {
-                        b.append(s == 0 ? "" : " AND ").append(o_).append(name);
+                        b.append(s == 0 ? "" : " AND ");
+                        appendAttributeName(name, b);
                         b.append(s < i ? '=' : (asc ? '<' : '>'));
                         b.append(paramPrefix).append(jpqlParamCount + 1 + s);
                     }
@@ -1921,9 +1938,8 @@ public class QueryInfo {
                             String name = getAttributeName(attribute, true);
 
                             q.append(first ? " " : ", ");
-                            if (name.charAt(name.length() - 1) != ')')
-                                q.append(o_);
-                            q.append(name).append("=");
+                            appendAttributeName(name, q);
+                            q.append("=");
                             first = false;
 
                             boolean withFunction = false;
@@ -2170,9 +2186,7 @@ public class QueryInfo {
                 // Specify columns without creating new instance
                 for (int i = 0; i < cols.length; i++) {
                     q.append(i == 0 ? "SELECT " : ", ");
-                    if (cols[i].charAt(cols[i].length() - 1) != ')')
-                        q.append(o_);
-                    q.append(cols[i]);
+                    appendAttributeName(cols[i], q);
                 }
             } else {
                 // Construct new instance from defined columns
@@ -2180,9 +2194,7 @@ public class QueryInfo {
                 for (int i = 0; i < cols.length; i++) {
                     if (i > 0)
                         q.append(", ");
-                    if (cols[i].charAt(cols[i].length() - 1) != ')')
-                        q.append(o_);
-                    q.append(cols[i]);
+                    appendAttributeName(cols[i], q);
                 }
                 q.append(')');
             }
@@ -2209,12 +2221,7 @@ public class QueryInfo {
         if (sort.ignoreCase())
             q.append("LOWER(");
 
-        if (propName.charAt(propName.length() - 1) == ')')
-            ; // id(o) or version(o) function
-        else
-            q.append(entityVar_);
-
-        q.append(propName);
+        appendAttributeName(propName, q);
 
         if (sort.ignoreCase())
             q.append(")");
@@ -2571,6 +2578,8 @@ public class QueryInfo {
                     Tr.debug(this, tc, "getCursorValues for " + loggable(entity),
                              accessors);
                 Object value = entity;
+                // TODO is it possible for accessors to be null here?
+                // Could sort.property() being VERSION(THIS) or some invalid name cause this?
                 for (Member accessor : accessors)
                     if (accessor instanceof Method)
                         value = ((Method) accessor).invoke(value);
@@ -3511,7 +3520,6 @@ public class QueryInfo {
                      countPages);
 
         String o = entityVar;
-        String o_ = entityVar_;
         StringBuilder q = null;
 
         if (methodTypeAnno instanceof Find || methodTypeAnno instanceof Update) {
@@ -4369,9 +4377,13 @@ public class QueryInfo {
         if (jpqlParamNames.isEmpty()) // positional parameters
             for (int i = 0; i < cursor.size(); i++) {
                 Object value = cursor.get(i);
-                if (entityInfo.idClassAttributeAccessors != null && entityInfo.idType.isInstance(value)) {
+                if (entityInfo.idClassAttributeAccessors != null &&
+                    entityInfo.idType.isInstance(value)) {
+                    // Expand ID(THIS) for composite IdClass into separate attributes
                     for (Member accessor : entityInfo.idClassAttributeAccessors.values()) {
-                        Object v = accessor instanceof Field ? ((Field) accessor).get(value) : ((Method) accessor).invoke(value);
+                        Object v = accessor instanceof Field //
+                                        ? ((Field) accessor).get(value) //
+                                        : ((Method) accessor).invoke(value);
                         if (++paramNum - jpqlParamCount > sorts.size())
                             cursorSizeMismatchError(cursor);
                         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
@@ -4392,13 +4404,18 @@ public class QueryInfo {
         else // named parameters
             for (int i = 0; i < cursor.size(); i++) {
                 Object value = cursor.get(i);
-                if (entityInfo.idClassAttributeAccessors != null && entityInfo.idType.isInstance(value)) {
+                if (entityInfo.idClassAttributeAccessors != null &&
+                    entityInfo.idType.isInstance(value)) {
+                    // Expand ID(THIS) for composite IdClass into separate attributes
                     for (Member accessor : entityInfo.idClassAttributeAccessors.values()) {
-                        Object v = accessor instanceof Field ? ((Field) accessor).get(value) : ((Method) accessor).invoke(value);
+                        Object v = accessor instanceof Field //
+                                        ? ((Field) accessor).get(value) //
+                                        : ((Method) accessor).invoke(value);
                         if (++paramNum - jpqlParamCount > sorts.size())
                             cursorSizeMismatchError(cursor);
                         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
-                            Tr.debug(this, tc, "set [cursor] :cursor" + paramNum + ' ' + value.getClass().getName() + "-->" +
+                            Tr.debug(this, tc, "set [cursor] :cursor" + paramNum +
+                                               ' ' + value.getClass().getName() + "-->" +
                                                (v == null ? null : v.getClass().getSimpleName()));
                         query.setParameter("cursor" + paramNum, v);
                     }
@@ -4472,7 +4489,8 @@ public class QueryInfo {
      * @return the combined list that the sort criteria was added to.
      */
     @Trivial
-    List<Sort<Object>> supplySorts(List<Sort<Object>> combined, Iterable<Sort<Object>> additional) {
+    List<Sort<Object>> supplySorts(List<Sort<Object>> combined,
+                                   Iterable<Sort<Object>> additional) {
         Iterator<Sort<Object>> addIt = additional.iterator();
         boolean hasIdClass = entityInfo.idClassAttributeAccessors != null;
         if (combined == null && addIt.hasNext())
@@ -4481,7 +4499,8 @@ public class QueryInfo {
             Sort<Object> sort = addIt.next();
             if (sort == null)
                 throw new IllegalArgumentException("Sort: null");
-            // TODO special IdClass handling should be unnecessary once 30093 is fixed
+            // IdClass is split up so that it can be possible to create a cursor
+            // that corresponds to sort criteria
             else if (hasIdClass && ID.equalsIgnoreCase(sort.property()))
                 for (String name : entityInfo.idClassAttributeAccessors.keySet())
                     combined.add(getWithAttributeName(getAttributeName(name, true), sort));
@@ -4501,14 +4520,16 @@ public class QueryInfo {
      * @return the combined list that the sort criteria was added to.
      */
     @Trivial
-    List<Sort<Object>> supplySorts(List<Sort<Object>> combined, @SuppressWarnings("unchecked") Sort<Object>... additional) {
+    List<Sort<Object>> supplySorts(List<Sort<Object>> combined,
+                                   @SuppressWarnings("unchecked") Sort<Object>... additional) {
         boolean hasIdClass = entityInfo.idClassAttributeAccessors != null;
         if (combined == null && additional.length > 0)
             combined = sorts == null ? new ArrayList<>() : new ArrayList<>(sorts);
         for (Sort<Object> sort : additional) {
             if (sort == null)
                 throw new IllegalArgumentException("Sort: null");
-            // TODO special IdClass handling should be unnecessary once 30093 is fixed
+            // IdClass is split up so that it can be possible to create a cursor
+            // that corresponds to sort criteria
             else if (hasIdClass && ID.equalsIgnoreCase(sort.property()))
                 for (String name : entityInfo.idClassAttributeAccessors.keySet())
                     combined.add(getWithAttributeName(getAttributeName(name, true), sort));

--- a/dev/io.openliberty.data.internal_fat/test-applications/DataTestApp/src/test/jakarta/data/web/DataTestServlet.java
+++ b/dev/io.openliberty.data.internal_fat/test-applications/DataTestApp/src/test/jakarta/data/web/DataTestServlet.java
@@ -3892,8 +3892,7 @@ public class DataTestServlet extends FATServlet {
      * Verify a repository method that supplies id(this) as the sort criteria
      * hard coded within a JDQL query.
      */
-    // TODO enable once #30093 is fixed
-    //@Test
+    @Test
     public void testOrderByIdFunction() {
         assertIterableEquals(List.of(19L, 17L, 13L, 11L, 7L, 5L, 3L, 2L),
                              primes.below(20L));

--- a/dev/io.openliberty.data.internal_fat/test-applications/DataTestApp/src/test/jakarta/data/web/Vehicles.java
+++ b/dev/io.openliberty.data.internal_fat/test-applications/DataTestApp/src/test/jakarta/data/web/Vehicles.java
@@ -66,8 +66,6 @@ public interface Vehicles {
 
     boolean updateByVinIdAddPrice(String vin, float priceIncrease);
 
-    // TODO switch to the following once #29893 is fixed
-    //@Query("WHERE LOWER(ID(THIS)) = ?1")
-    @Query("WHERE LOWER(vinId) = ?1")
+    @Query("WHERE LOWER(ID(THIS)) = ?1")
     Optional<Vehicle> withVINLowerCase(String lowerCaseVIN);
 }

--- a/dev/io.openliberty.data.internal_fat_errorpaths/test-applications/DataErrPathsTestApp/src/test/jakarta/data/errpaths/web/DataErrPathsTestServlet.java
+++ b/dev/io.openliberty.data.internal_fat_errorpaths/test-applications/DataErrPathsTestApp/src/test/jakarta/data/errpaths/web/DataErrPathsTestServlet.java
@@ -470,9 +470,8 @@ public class DataErrPathsTestServlet extends FATServlet {
     }
 
     /**
-     * Tests an error path where a repository method attempts to return a subset of
-     * an entity as a record where the record component names do not all match the
-     * entity attribute names.
+     * Tests an error path where a repository method attempts to remove an entity
+     * but return it as a record instead.
      */
     @Test
     public void testRemoveAsSubsetOfEntity() {

--- a/dev/io.openliberty.data.internal_fat_exp/test-applications/DataExperimentalWeb/src/test/jakarta/data/experimental/web/DataExperimentalServlet.java
+++ b/dev/io.openliberty.data.internal_fat_exp/test-applications/DataExperimentalWeb/src/test/jakarta/data/experimental/web/DataExperimentalServlet.java
@@ -524,23 +524,6 @@ public class DataExperimentalServlet extends FATServlet {
 
     /**
      * Repository method performing a parameter-based query on a compound entity Id which is an IdClass,
-     * where the method parameter is annotated with By.
-     */
-    // TODO enable once #29893 is fixed. EclipseLink rejects LOWER(id(o)) with:
-    // The encapsulated expression is not a valid expression.
-    //@Test
-    public void testIdClassFindByAnnotatedParameter() {
-
-        assertEquals(List.of("Springfield Massachusetts",
-                             "Rochester Minnesota",
-                             "Kansas City Missouri"),
-                     towns.largerThan(100000, TownId.of("springfield", "missouri"), "M%s")
-                                     .map(c -> c.name + ' ' + c.stateName)
-                                     .collect(Collectors.toList()));
-    }
-
-    /**
-     * Repository method performing a parameter-based query on a compound entity Id which is an IdClass,
      * without annotating the method parameter.
      */
     // TODO enable once #29073 is fixed
@@ -721,6 +704,20 @@ public class DataExperimentalServlet extends FATServlet {
 
         assertEquals(List.of("seventeen"),
                      primes.withRomanNumeralSuffixAndWithoutNameSuffix("VII", "seven", 50));
+    }
+
+    /**
+     * Test the NotIgnoreCase enumerated value for a parameter that is
+     * annotated with the By annotation.
+     */
+    @Test
+    public void testNotIgnoreCase() {
+
+        assertEquals(List.of("Rochester Minnesota",
+                             "Kansas City Missouri"),
+                     towns.largerThan(100000, "springfield", "M%s")
+                                     .map(c -> c.name + ' ' + c.stateName)
+                                     .collect(Collectors.toList()));
     }
 
     /**

--- a/dev/io.openliberty.data.internal_fat_exp/test-applications/DataExperimentalWeb/src/test/jakarta/data/experimental/web/DataExperimentalServlet.java
+++ b/dev/io.openliberty.data.internal_fat_exp/test-applications/DataExperimentalWeb/src/test/jakarta/data/experimental/web/DataExperimentalServlet.java
@@ -585,11 +585,10 @@ public class DataExperimentalServlet extends FATServlet {
 
     /**
      * Use cursor-based pagination with the OrderBy annotation on a composite id
-     * that is defined by an IdClass attribute. Also use named parameters, which
-     * means the cursor portion of the query will also need to use named parameters.
+     * that is defined by an IdClass attribute.
      */
     @Test
-    public void testIdClassOrderByAnnotationWithCursorPaginationAndNamedParameters() {
+    public void testIdClassOrderByAnnotationWithCursorPaginations() {
         PageRequest pagination = PageRequest.ofSize(2);
 
         CursoredPage<Town> page1 = towns.sizedWithin(100000, 1000000, pagination);

--- a/dev/io.openliberty.data.internal_fat_exp/test-applications/DataExperimentalWeb/src/test/jakarta/data/experimental/web/Towns.java
+++ b/dev/io.openliberty.data.internal_fat_exp/test-applications/DataExperimentalWeb/src/test/jakarta/data/experimental/web/Towns.java
@@ -85,7 +85,7 @@ public interface Towns {
     @OrderBy("stateName")
     @OrderBy("name")
     Stream<Town> largerThan(@By("population") @Is(GreaterThan) int minPopulation,
-                            @By(ID) @Is(NotIgnoreCase) TownId exceptFor,
+                            @By("name") @Is(NotIgnoreCase) String cityToExclude,
                             @By("stateName") @Is(Prefixed) String statePattern);
 
     @Update

--- a/dev/io.openliberty.data.internal_fat_jpa/test-applications/DataJPATestApp/src/test/jakarta/data/jpa/web/Cities.java
+++ b/dev/io.openliberty.data.internal_fat_jpa/test-applications/DataJPATestApp/src/test/jakarta/data/jpa/web/Cities.java
@@ -131,6 +131,12 @@ public interface Cities {
     @Save
     City save(City c);
 
+    @Query("WHERE population < :maxPopulation OR name <> :nameToExclude")
+    @OrderBy(ID)
+    CursoredPage<City> smallerThanOrNotNamed(int maxPopulation,
+                                             String nameToExclude,
+                                             PageRequest pageReq);
+
     @Find
     @OrderBy("stateName")
     Stream<City> withNameOf(String name);

--- a/dev/io.openliberty.data.internal_fat_jpa/test-applications/DataJPATestApp/src/test/jakarta/data/jpa/web/DataJPATestServlet.java
+++ b/dev/io.openliberty.data.internal_fat_jpa/test-applications/DataJPATestApp/src/test/jakarta/data/jpa/web/DataJPATestServlet.java
@@ -707,6 +707,42 @@ public class DataJPATestServlet extends FATServlet {
     }
 
     /**
+     * Test that a query with named parameters can be used for pagination
+     * where the cursor is an IdClass value.
+     */
+    @Test
+    public void testCursoredPagesWithNamedParametersAndIdClass() {
+        PageRequest page1req = PageRequest
+                        .ofSize(3)
+                        .afterCursor(Cursor.forKey(CityId.of("Indianapolis",
+                                                             "Indiana")));
+
+        CursoredPage<City> page1 = cities.smallerThanOrNotNamed(100000,
+                                                                "Springfield",
+                                                                page1req);
+
+        assertEquals(List.of("Kansas City in Kansas",
+                             "Kansas City in Missouri",
+                             "Rochester in Minnesota"),
+                     page1.stream()
+                                     .map(c -> c.name + " in " + c.stateName)
+                                     .collect(Collectors.toList()));
+
+        CursoredPage<City> page2 = cities.smallerThanOrNotNamed(100000,
+                                                                "Springfield",
+                                                                page1.nextPageRequest());
+
+        assertEquals(List.of("Rochester in New York",
+                             "Springfield in Ohio",
+                             "Springfield in Oregon"),
+                     page2.stream()
+                                     .map(c -> c.name + " in " + c.stateName)
+                                     .collect(Collectors.toList()));
+
+        assertEquals(false, page2.hasNext());
+    }
+
+    /**
      * Verify that an EntityManager can be obtained for a repository and used to perform database operations.
      */
     @Test

